src/domain/common/lifecycle/lifecycle.service.ts

```diff
@@ -82,14 +92,20 @@ export class LifecycleService {
     this.logger.verbose?.(
       `Lifecycle (id: ${
         lifecycle.id
-      }) event '${eventName}: from state '${startState}' to state '${machineService.state.value.toString()}', parentID: ${parentID}`,
+      }) event '${eventName}: from state '${restoredState.value.toString()}' to state '${machineService.state.value.toString()}', parentID: ${parentID}`,
       LogContext.LIFECYCLE
     );
 
-    const newStateStr = JSON.stringify(machineService.state);
+    const stateToHydrate = interpretedLifecycle.state;
+    // Note: https://github.com/statelyai/xstate/discussions/1757
+    // restoring state has the hydrated actions, which unless removed will be executed again
+    stateToHydrate.actions = [];
+    const newStateStr = JSON.stringify(stateToHydrate);
+
+    // Todo: do not stop as this triggers an exit action from the last state.
+    //machineService.stop();
```

Comment:
So I guess my hunch that the machine needs to be singleton was on the right path...

Reply:
It is not that the machine needs to be a singleton, there can be many machines in parallel. It is to do with the way the state is hydrated, I think they will change it.